### PR TITLE
Limit state tolerance

### LINKED ISSUE #2559: FORM / SORM needs modification to avoid bad estimation of probability in case of non convergence of optimization
### What is the idea?

Hi,

With the new feature of setting ignoreFailure of the optimization algorithm to `True` inside FORM / SORM, there is a possibility that the optimization algorithm does not converge in the most probable failure point estimation process embedded in FORM / SORM. 

In case of bad convergence, the equality constraint satisfaction about being on the limit state function is not ensured. In this case, the failure probability estimation is not possible.

One test should be included in FORM / SORM algorithms to check the constraint satisfaction after the optimization algorithm run and retrieve an exception if the constraint is not satisfied.

### Why is this needed?

_No response_

### Additional Context

To solve this issue, one may add a `limitStateTolerance` attribute to let the user choose the tolerance on the constraint satisfaction, and implement a test after the optimization algorithm run to check the validity of the results.

### PR DESCRIPTION
**Pull request by m-balesdent**
Hello,

Closes #2559 .

This PR aims to fix the issue encountered for FORM /SORM when `setCheckStatus` is set to `False` for the optimization algorithm  such as Cobyla or when algorithms derived from other libraries like NLopt or Pagmo are used. Indeed, in that case, the convergence of the optimizer is not checked so it is not ensured that the design point is on the limit state.

This PR implements a test in order to check if the result found by the optimization algorithm satisfies the equality constraint dealing with the limit state function. This test depends on a parameter, named `limitStateTolerance` that can be defined by the user.  Default value of 0.01 is defined in the ResourceMap. Accessors have been created.

The test is implemented in the base  class `Analytical` and all the derived classes (`FORM`, `multiFORM`, `SORM` and `SystemFORM`) have been updated (constructors).

### CODE REVIEW

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx [hunk 1]
```diff
@@ -57,8 +57,12 @@ public:
   Analytical(const OptimizationAlgorithm & solver,
              const RandomVector & compositeEvent,
              const Point & physicalStartingPoint);
-
-
+             
+  Analytical(const OptimizationAlgorithm & solver,
+             const RandomVector & compositeEvent,
+             const Point & physicalStartingPoint,
+             const Scalar & limitStateTolerance);
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx:64]: we dont want to add new ctors each time parameters are added, instead it can be already set with the accessor you added

Comment (m-balesdent) [reply lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx:64]: So you recommend just to access it by setLimitStateTolerance ?

Comment (jschueller) [reply lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx:64]: yes

File: lib/src/Base/Common/ResourceMap.cxx [hunk 2]
```diff
@@ -1380,6 +1380,9 @@ void ResourceMap::loadDefaultConfiguration()
   addAsUnsignedInteger("AdaptiveDirectionalStratification-DefaultMaximumStratificationDimension", 3);
   addAsUnsignedInteger("AdaptiveDirectionalStratification-DefaultNumberOfSteps", 2);
 
+  // Analytical parameters //
+  addAsScalar("Analytical-defaultLimitStateTolerance", 0.01);
```
Comment (jschueller) [inline lib/src/Base/Common/ResourceMap.cxx:1384]: must start with uppercase

Comment (jschueller) [reply lib/src/Base/Common/ResourceMap.cxx:1384]: also it must be added to openturns.conf.in

File: python/src/Analytical_doc.i.in [hunk 3]
```diff
@@ -178,3 +183,27 @@ Notes
...
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::Analytical::getLimitStateTolerance
+"Accessor to limitStateTolerance parameter.
+
+Returns
+-------
+limitStateTolerance : float
+    Tolerance on limit state to assess if the found design point can be used 
+    to estimate the failure probability. Default value is defined in :class:`~openturns.ResourceMap` of  class :class:`~openturns.Analytical`. "
```
Comment (jschueller) [inline python/src/Analytical_doc.i.in:196]: better mention the full key name than just the class

File: python/src/Analytical_doc.i.in [hunk 4]
```diff
@@ -178,3 +183,27 @@ Notes
...
+
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::Analytical::setLimitStateTolerance
+"Accessor to limitStateTolerance parameter.
+
+Returns
```
Comment (jschueller) [inline python/src/Analytical_doc.i.in:203]: Parameters

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/FORM.hxx [hunk 5]
```diff
@@ -45,12 +45,12 @@ public:
   /** Default constructor */
   FORM();
 
-  /** Constructor with parameters */
+  /** Constructors with parameters */
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/Analytical/openturns/FORM.hxx:48]: should be reverted

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx [hunk 6]
```diff
@@ -103,12 +110,12 @@ protected:
   friend class Factory<Analytical>;
 
 private:
-
   OptimizationAlgorithm nearestPointAlgorithm_;
   RandomVector event_;
   Point physicalStartingPoint_;
   AnalyticalResult result_;
-
+  Scalar limitStateTolerance_;
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx:117]:
```suggestion
  Scalar limitStateTolerance_ = 0.0;
```

File: python/src/Analytical_doc.i.in (hunk 3 above)
Comment (m-balesdent) [reply python/src/Analytical_doc.i.in:196]: fixed

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/FORM.hxx (hunk 5 above)
Comment (m-balesdent) [reply lib/src/Uncertainty/Algorithm/Analytical/openturns/FORM.hxx]: fixed

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx (hunk 6 above)
Comment (m-balesdent) [reply lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx]: fixed

File: lib/src/Base/Common/ResourceMap.cxx (hunk 2 above)
Comment (m-balesdent) [reply lib/src/Base/Common/ResourceMap.cxx]: fixed

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx [hunk 7]
```diff
@@ -131,6 +146,18 @@ void Analytical::run()
...
+  
+  /* check is result is valid */
+  Point physicalSpaceDesignPoint(event_.getImplementation()->getAntecedent().getDistribution().getInverseIsoProbabilisticTransformation().operator()(standardSpaceDesignPoint));
+  
+  Point valuePhysicalSpaceDesignPoint(event_.getImplementation()->getFunction().operator()(physicalSpaceDesignPoint));
+  
+  Scalar residual = (valuePhysicalSpaceDesignPoint[0]-event_.getThreshold())*(valuePhysicalSpaceDesignPoint[0]-event_.getThreshold());  
+  
+  if (residual > limitStateTolerance_*limitStateTolerance_)
+  throw Exception(HERE) << "Obtained design point is not on the limit state : its image by the limit state function is " << valuePhysicalSpaceDesignPoint[0] << " that is uncompatible with the threshold : " << event_.getThreshold() << " considering the limit state tolerance "<< limitStateTolerance_;
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx:158]: Add a space befor the last <<

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/FORM.hxx [hunk 8]
```diff
@@ -50,7 +50,7 @@ public:
        const RandomVector & event,
        const Point & physicalStartingPoint);
 
-
+
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Analytical/openturns/FORM.hxx:53]: Remove the blanks here

File: python/test/t_FORM_interval.py [hunk 9]
```diff
@@ -62,6 +62,8 @@
     ot.RandomGenerator.SetSeed(0)
     # algo = getattr(openturns, algoName)(event)
     algo = ot.FORM(ot.Cobyla(), event, X.getMean())
+    algo.setLimitStateTolerance(2.)
```
Comment (regislebrun) [inline python/test/t_FORM_interval.py:65]: ??? Strange value for a tolerance. Maybe the case where the event is always true should be treated separately?
BTW prefer 2.0 to 2., it makes it clearer that you use a floating point value.

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/FORM.hxx (hunk 8 above)
Comment (m-balesdent) [reply lib/src/Uncertainty/Algorithm/Analytical/openturns/FORM.hxx:53]: fixed, thanks.

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx (hunk 7 above)
Comment (m-balesdent) [reply lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx]: fixed, thanks.

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx [hunk 10]
```diff
@@ -131,6 +146,18 @@ void Analytical::run()
...
   result_.setIsStandardPointOriginInFailureSpace(event_.getOperator().compare(value[0], event_.getThreshold()));
+  
+  /* check is result is valid */
+  Point physicalSpaceDesignPoint(event_.getImplementation()->getAntecedent().getDistribution().getInverseIsoProbabilisticTransformation().operator()(standardSpaceDesignPoint));
+  
+  Point valuePhysicalSpaceDesignPoint(event_.getImplementation()->getFunction().operator()(physicalSpaceDesignPoint));
+  
+  Scalar residual = (valuePhysicalSpaceDesignPoint[0]-event_.getThreshold())*(valuePhysicalSpaceDesignPoint[0]-event_.getThreshold());  
+  
+  if (residual > limitStateTolerance_*limitStateTolerance_)
```
[the remark this answers is not in the document]
Comment (m-balesdent) [reply lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx]: fixed, thanks.

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx [hunk 11]
```diff
@@ -131,6 +146,18 @@ void Analytical::run()
   Point value(standardEvent.getImplementation()->getFunction().operator()(origin));
 
   result_.setIsStandardPointOriginInFailureSpace(event_.getOperator().compare(value[0], event_.getThreshold()));
+  
+  /* check is result is valid */
+  Point physicalSpaceDesignPoint(event_.getImplementation()->getAntecedent().getDistribution().getInverseIsoProbabilisticTransformation().operator()(standardSpaceDesignPoint));
+  
+  Point valuePhysicalSpaceDesignPoint(event_.getImplementation()->getFunction().operator()(physicalSpaceDesignPoint));
+  
+  Scalar residual = (valuePhysicalSpaceDesignPoint[0]-event_.getThreshold())*(valuePhysicalSpaceDesignPoint[0]-event_.getThreshold());
```
[the remark this answers is not in the document]
Comment (m-balesdent) [reply lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx]: fixed, thanks.

File: lib/etc/openturns.conf.in [hunk 12]
```diff
@@ -750,6 +750,10 @@
   <AdaptiveDirectionalStratification-DefaultMaximumStratificationDimension value_int="3" />
   <AdaptiveDirectionalStratification-DefaultNumberOfSteps                  value_int="2" />
 
+
+  <!-- OT::AnalyticalResult parameters -->
```
[the remark this answers is not in the document]
Comment (m-balesdent) [reply lib/etc/openturns.conf.in]: fixed, thanks.

File: python/test/t_FORM_interval.py (hunk 9 above)
Comment (m-balesdent) [reply python/test/t_FORM_interval.py]: In that tests, FORM does not converge to the probability, that's why the value found by analytical is not compatible with the limitStateTolerance.

Formating typo has been fixed.

File: lib/src/Uncertainty/Algorithm/Analytical/FORM.cxx [hunk 13]
```diff
@@ -48,6 +48,7 @@ FORM::FORM(const OptimizationAlgorithm & nearestPointAlgorithm,
   // Nothing to do
 }
 
+
```
Comment (josephmure) [inline lib/src/Uncertainty/Algorithm/Analytical/FORM.cxx:51]: Why is this line added (here and in other files)?

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx [hunk 14]
```diff
@@ -103,12 +110,12 @@ protected:
   friend class Factory<Analytical>;
 
 private:
-
   OptimizationAlgorithm nearestPointAlgorithm_;
   RandomVector event_;
   Point physicalStartingPoint_;
   AnalyticalResult result_;
-
+  Scalar limitStateTolerance_ = 0.0;
```
Comment (josephmure) [inline lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx:117]: This is odd: if the default value is defined in the ResourceMap (to 0.01), you don't need to define a different default here.

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/FORM.hxx (hunk 8 above)
Comment (josephmure) [inline lib/src/Uncertainty/Algorithm/Analytical/openturns/FORM.hxx:53]: There are still blanks here (cf. Régis' earlier remark).

File: python/src/Analytical_doc.i.in [hunk 15]
```diff
@@ -178,3 +183,27 @@ Notes
...
+
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::Analytical::getLimitStateTolerance
+"Accessor to the limit state tolerance.
+
+Returns
+-------
+limitStateTolerance : float
+    Tolerance on limit state to assess if the found design point can be used
```
Comment (josephmure) [inline python/src/Analytical_doc.i.in:195]: Maybe specify that this is an absolute tolerance.

File: python/src/Analytical_doc.i.in [hunk 16]
```diff
@@ -178,3 +183,27 @@ Notes
 Performs the research of the design point and creates a
 :class:`~openturns.AnalyticalResult`, the structure result which is
 accessible with the method :meth:`getAnalyticalResult`."
+
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::Analytical::getLimitStateTolerance
```
Comment (josephmure) [inline python/src/Analytical_doc.i.in:189]: I think Julien's earlier remarks were not taken into account, despite appearing as resolved. Maybe the relevant commits were lost?

**Review by josephmure — CHANGES_REQUESTED** [review]
Good job, a few details need to be tweaked.

File: python/src/Analytical_doc.i.in (hunk 16 above)
Comment (m-balesdent) [reply python/src/Analytical_doc.i.in:189]: For me, the remark is about adding a new constructor of the class depending on this parameter. I removed the ctor.

Comment (josephmure) [reply python/src/Analytical_doc.i.in:189]: I meant the remarks about stating the [full name of the ResourceMap key](https://github.com/openturns/openturns/pull/2561/files#r1513021079) and about [replacing "Returns" with "Parameters"](https://github.com/openturns/openturns/pull/2561/files#r1513021873) in the "set" accessor.

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx [hunk 17]
```diff
@@ -131,6 +146,18 @@ void Analytical::run()
...
+  
+  /* check if result is valid */
+  Point physicalSpaceDesignPoint(event_.getImplementation()->getAntecedent().getDistribution().getInverseIsoProbabilisticTransformation().operator()(standardSpaceDesignPoint));
+  
+  Point valuePhysicalSpaceDesignPoint(event_.getImplementation()->getFunction().operator()(physicalSpaceDesignPoint));
+  
+  const Scalar residual = std::abs(valuePhysicalSpaceDesignPoint[0] - event_.getThreshold()); 
+  
+  if (!(residual <= limitStateTolerance_))
+  throw Exception(HERE) << "Obtained design point is not on the limit state : its image by the limit state function is " << valuePhysicalSpaceDesignPoint[0] << ", which is incompatible with the threshold : " << event_.getThreshold() << " considering the limit state tolerance : "<< limitStateTolerance_;
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx:158]: Citation (from [learnenglish.de](https://www.learnenglish.de/punctuation/punctuationtext.html)): "You do not put a space before a colon, but you do need a space after one".
```suggestion
  throw Exception(HERE) << "Obtained design point is not on the limit state: its image by the limit state function is " << valuePhysicalSpaceDesignPoint[0] << ", which is incompatible with the threshold: " << event_.getThreshold() << " considering the limit state tolerance: "<< limitStateTolerance_;
```

File: python/src/Analytical_doc.i.in [hunk 18]
```diff
@@ -80,6 +80,11 @@ The result of the optimization problem is recoverable thanks to the method
 The unicity and the strongness of the design point can be checked thanks to the
 :class:`Strong Maximum Test <openturns.StrongMaximumTest>`.
 
+A tolerance on the limit state (in the physical space) is implemented to check if the found design point can be used to estimate the failure probability, that is if
+:math:`g(\vect{x}^*,\vect{d}) = 0`.
```
Comment (mbaudin47) [inline python/src/Analytical_doc.i.in:84]: I think the message should present the exact mathematical equation.
```suggestion
An absolute tolerance on the limit state (in the physical space) is implemented to check if the found design point can be used to estimate the failure probability, that is if:

.. math::
    \left| \model(\vect{x}^*) - s\right|\leq \epsilon_t

where :math:`\vect{x}^*` is the design point, :math:`s` is the threshold and :math:`\epsilon_t`
is the absolute tolerance.
```

File: lib/src/Base/Common/ResourceMap.cxx [hunk 19]
```diff
@@ -1380,6 +1380,9 @@ void ResourceMap::loadDefaultConfiguration()
   addAsUnsignedInteger("AdaptiveDirectionalStratification-DefaultMaximumStratificationDimension", 3);
   addAsUnsignedInteger("AdaptiveDirectionalStratification-DefaultNumberOfSteps", 2);
 
+  // Analytical parameters //
+  addAsScalar("Analytical-DefaultLimitStateTolerance", 0.01);
```
Comment (mbaudin47) [inline lib/src/Base/Common/ResourceMap.cxx:1384]: I do not know a perfect method to set the default tolerance, but this seems quite loose to me. Would it be relevant to be more stringent, e.g. $10^{-5}$?

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx [hunk 20]
```diff
@@ -131,6 +146,18 @@ void Analytical::run()
...
   result_.setIsStandardPointOriginInFailureSpace(event_.getOperator().compare(value[0], event_.getThreshold()));
+  
+  /* check if result is valid */
+  Point physicalSpaceDesignPoint(event_.getImplementation()->getAntecedent().getDistribution().getInverseIsoProbabilisticTransformation().operator()(standardSpaceDesignPoint));
+  
+  Point valuePhysicalSpaceDesignPoint(event_.getImplementation()->getFunction().operator()(physicalSpaceDesignPoint));
+  
+  const Scalar residual = std::abs(valuePhysicalSpaceDesignPoint[0] - event_.getThreshold()); 
+  
+  if (!(residual <= limitStateTolerance_))
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx:157]: - It seems to me that, if the value of the threshold has a very large magnitude or is very close to zero, it might be difficult to chose a relevant value of the tolerance. For example, if the threshold is a pressure, the order of magnitude might be close to $10^9$. In this case, an absolute tolerance equal to 0.01 is _extremely_ stringent.
- We cannot use a relative tolerance neither, because the user-defined threshold might be zero.
- One way to solve the problem is to use a mixed absolute-relative criteria. Introducing a new `limitStateRelativeTolerance_` parameter, this goes like:

```suggestion
  if (!(std::abs(valuePhysicalSpaceDesignPoint[0] - event_.getThreshold()) <= limitStateAbsoluteTolerance_ + limitStateRelativeTolerance_ * std::abs(event_.getThreshold())))
```

**Review by mbaudin47 — CHANGES_REQUESTED** [review]
This seems a very necessary PR: thank you for sharing this. The current code, however, might be discussed further to check that the current implementation is the best possible option.

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx (hunk 20 above)
Comment (m-balesdent) [reply lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx:157]: I agree with you about the fact that the threshold is problem-dependent. IMO, in most of the real world problems, the tolerance is given by expert so it has to be tuned before using the algorithm. I'm not convinced that mixing the absolute and relative tolerance in a composite criterion will be the most easy to understand for the user. Maybe we can discuss in depth of this during the next technical committee?

File: lib/src/Base/Common/ResourceMap.cxx (hunk 19 above)
Comment (m-balesdent) [reply lib/src/Base/Common/ResourceMap.cxx:1384]: This value is very dependent of the optimization problem formulation and physics involved. 
I agree with you about the need to name it "Analytical-DefaultLimitStateAbsoluteTolerance".

**Comment by m-balesdent** [comment]
@mbaudin47 there is also an additional complexity. The optimization in FORM is performed in standard space. So the tolerance defined for the embedded optimization algorithm is difficult to be handled by the user since an isoprobabilistic transformation is involved. I think the absolute tolerance should be defined in the physical space for user comfort.

**Review by regislebrun — APPROVED** [review]
Good to go once the issues raised by the other reviewers are addressed.

**Comment by regislebrun** [comment]
Any news from this PR? It solves an important bug, so it should definitely be part of the next release!

**Comment by m-balesdent** [comment]
> Any news from this PR? It solves an important bug, so it should definitely be part of the next release!

The way to define the tolerance has not reached an agreement yet. Michael proposes to implement a combination of relative and absolute tolerances, but IMO it is quite dangerous because for reliability purpose, the user has to be sure about which value of the threshold is considered (in physical space). This value is something that is often a specification of the study. 

So I propose to have a debate on next technical commitee in order to have a consensus on this important feature. I will next implement the solution that will have been chosen.

**Comment by mbaudin47** [comment]
> So I propose to have a debate on next technical committee in order to have a consensus on this important feature. I will next implement the solution that will have been chosen.

I suggest to prepare the discussion by creating one Python script comparing the type of tolerance on the residual of the threshold:
- one based on a single absolute tolerance,
- one based on a mixed absolute / relative tolerance.

This example could be based on meaningful tolerance values such as $10^{-4}$ and various threshold values such as $0$ (!), $10$, ..., $10^9$, ..., $10^{16}$. For each case, we may compare the result of the two conditions we suggest to implement with respect to the expected result.

**Comment by m-balesdent** [comment]
> > So I propose to have a debate on next technical committee in order to have a consensus on this important feature. I will next implement the solution that will have been chosen.
> 
> I suggest to prepare the discussion by creating one Python script comparing the type of tolerance on the residual of the threshold:
> 
>     * one based on a single absolute tolerance,
> 
>     * one based on a mixed absolute / relative tolerance.
> 
> 
> This example could be based on meaningful tolerance values such as 10−4 and various threshold values such as 0 (!), 10, ..., 109, ..., 1016. For each case, we may compare the result of the two conditions we suggest to implement with respect to the expected result.

I think this is really dependent on the test case that is considered. IMO, asking the user to define a relative tolerance with respect to a reliability problem threshold is quite complicated, especially when the threshold is 0.

**Comment by m-balesdent** [comment]
Decision taken at the technical committe : get limit tolerance from optimization algorithm constraint error  and avoid addtional resource map key.

**Comment by m-balesdent** [comment]
Decision taken at the technical committe

> @jschueller following the last commit, it seems that cobyla does not converge exactly at the value indicated at MaximumConstraintError even if the test passes.
> 
> For example in the example that fails on compilation, value of error is actually -1.00044e-10 whereas the threshold is 1e-10. Consequently, it causes crashes of FORM algorithm.
> 
> If I put 1e-5 as MaximumConstraintError , the calculated error at the end of convergence will be 1.000x e-5 that is above the threshold.
> 
> Do you have an idea to force threshold ?

--> to prevent numerical instabilities in Cobyla convergence, a multiplicative factor of 1.1 has to be added in the test.

**Comment by m-balesdent** [comment]
@jschueller apparently, nlopt is not included in mindeps so the tests fail. Is it a way to overcome this limitation? The tests pass on the other platforms..

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx [hunk 21]
```diff
@@ -131,6 +131,19 @@ void Analytical::run()
...
+  Point physicalSpaceDesignPoint(event_.getImplementation()->getAntecedent().getDistribution().getInverseIsoProbabilisticTransformation().operator()(standardSpaceDesignPoint));
+  
+  Point valuePhysicalSpaceDesignPoint(event_.getImplementation()->getFunction().operator()(physicalSpaceDesignPoint));
+  
+  const Scalar residual = std::abs(valuePhysicalSpaceDesignPoint[0] - event_.getThreshold()); 
+  
+  const Scalar limitStateTolerance = nearestPointAlgorithm_.getMaximumConstraintError();
+  
+  if (!(residual <= 1.1 * limitStateTolerance)) // 1.1 is added to prevent numerical approximation made in Cobyla for constraint satisfaction tolerance.
+  throw Exception(HERE) << "Obtained design point is not on the limit state: its image by the limit state function is " << valuePhysicalSpaceDesignPoint[0] << ", which is incompatible with the threshold: " << event_.getThreshold() << " considering the limit state tolerance of the optimization algorithm: "<< limitStateTolerance;
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx:145]: Shouldn't this line be indented so that it is a part of the `if`?

File: python/test/t_FORM_interval.py [hunk 22]
```diff
@@ -61,7 +61,10 @@
 
     ot.RandomGenerator.SetSeed(0)
     # algo = getattr(openturns, algoName)(event)
-    algo = ot.FORM(ot.Cobyla(), event, X.getMean())
+    algoOptim = ot.Cobyla()
+    algoOptim.setMaximumConstraintError(1)
```
Comment (mbaudin47) [inline python/test/t_FORM_interval.py:65]: I am sorry, but I do not understand this line. Why disabling the maximum error constraint in this particular unit test?

File: python/test/t_Waarts_concave.py [hunk 23]
```diff
@@ -41,24 +41,6 @@
 
 #
 # Calculation
-# ot.
-#
-# FORM/SORM Cobyla
-myCobyla = ot.Cobyla()
```
Comment (mbaudin47) [inline python/test/t_Waarts_concave.py:47]: I guess that this block of unit test fails now that you check that the threshold is accurately reached. Am I correct? If yes, this means that Cobyla does not perform well. In this case, it might be relevant to create an issue that point this particular problem. This will prevent from forgetting this issue.

**Review by mbaudin47 — CHANGES_REQUESTED** [review]

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx (hunk 21 above)
Comment (m-balesdent) [reply lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx:145]: yes you're right!

File: python/test/t_FORM_interval.py (hunk 22 above)
Comment (m-balesdent) [reply python/test/t_FORM_interval.py:65]: In this test case, I think that for several domains, there is no solution that could reach the limit state. I tried different optimization algorithms (abdorackwitz, nlopt), but noone works.

For example, the fourth problem to be solved is : 

`class=ThresholdEventImplementation antecedent=class=CompositeRandomVector function=class=Function name=Unnamed implementation=class=FunctionImplementation name=Unnamed description=[x0,x1,x2,x3,y0] evaluationImplementation=class=SymbolicEvaluation name=Unnamed inputVariablesNames=[x0,x1,x2,x3] outputVariablesNames=[y0] formulas=[0.0] gradientImplementation=class=SymbolicGradient name=Unnamed evaluation=class=SymbolicEvaluation name=Unnamed inputVariablesNames=[x0,x1,x2,x3] outputVariablesNames=[y0] formulas=[0.0] hessianImplementation=class=SymbolicHessian name=Unnamed evaluation=class=SymbolicEvaluation name=Unnamed inputVariablesNames=[x0,x1,x2,x3] outputVariablesNames=[y0] formulas=[0.0] antecedent=class=UsualRandomVector distribution=class=Normal name=Normal dimension=4 mean=class=Point name=Unnamed dimension=4 values=[0,0,0,0] sigma=class=Point name=Unnamed dimension=4 values=[1,1,1,1] correlationMatrix=class=CorrelationMatrix dimension=4 implementation=class=MatrixImplementation name=Unnamed rows=4 columns=4 values=[1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1] operator=class=Less name=Unnamed threshold=1`

You will see that formula is `0.0` ! So it is non solvable.

File: python/test/t_Waarts_concave.py (hunk 23 above)
Comment (m-balesdent) [reply python/test/t_Waarts_concave.py:47]: Indeed, in this case (and other ones), Cobyla is not able to reach convergence. NLopt Cobyla performs quite well. I think this is not necessarily an issue. Depending on the problem specifications, several optimization algorithms could not find a solution.

File: python/test/t_FORM_interval.py (hunk 22 above)
Comment (mbaudin47) [reply python/test/t_FORM_interval.py:65]: This unit test is poorly designed for several reasons:
1) There is no particular reason to check a list of intervals by making variations on the finiteness of the bounds. If the algorithm works for one single interval, it will work for any. The only difference is whether the bounds are correctly taken into account when we check that a given point is within the bounds of an interval: this check is done within a unit test of the Interval class.
2) There is no comparison with a known probability value. We only check that the algorithm runs.

Now I understand what you did. Given that the there are unsolvable problems in the list, setting the tolerance to 1 essentially disables the check that you introduced. This is almost equivalent to disabling the whole unit test (but we still check that the constructor builds).

From my point of view, this unit test reveals that the check you introduce is interesting. This is why I validate this PR. There is, however, a work to be done on these unit tests so that relevant tests are chosen and correct reference probability values are used.

**Review by mbaudin47 — APPROVED** [review]
This PR adds a very significant value of the library: thank you for pointing this.

Furthermore, it reveals that:
- Cobyla can stop for reasons that are not understood at this point,
- many of the unit tests of the FORM class are poorly implemented.

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx (hunk 21 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx]: At this stage, launching the `run()` statement will produce an exception, whatever the residuals.

**Comment by m-balesdent** [comment]
@jschueller I think I answered all the issues for this PR. Maybe it can be merged ?

**Comment by jschueller** [comment]
hi, I think that we could use the computed constraint error from the OptimizationResult instead of recomputing it manually
else the ResourceMap key is unused, please drop it
also could you rebase to master ?

File: lib/etc/openturns.conf.in [hunk 24]
```diff
@@ -750,6 +750,10 @@
   <AdaptiveDirectionalStratification-DefaultMaximumStratificationDimension value_int="3" />
   <AdaptiveDirectionalStratification-DefaultNumberOfSteps                  value_int="2" />
 
+
+  <!-- OT::Analytical parameters -->
+  <Analytical-DefaultLimitStateTolerance      value_float="0.01"           />
```
Comment (jschueller) [inline lib/etc/openturns.conf.in:755]: this can be drop if not used

File: python/src/FORM_doc.i.in [hunk 25]
```diff
@@ -3,6 +3,10 @@
 
 Refer to :ref:`form_approximation`.
 
+Available constructors:
+    FORM(*nearestPointAlgorithm, event, physicalStartingPoint*)
```
Comment (jschueller) [inline python/src/FORM_doc.i.in:7]: we dont need a list of available ctors when there is only one available: the list of parameters is enough

File: lib/etc/openturns.conf.in [hunk 26]
```diff
@@ -749,7 +749,7 @@
   <AdaptiveDirectionalStratification-DefaultGamma                          value_float="0.5"         />
   <AdaptiveDirectionalStratification-DefaultMaximumStratificationDimension value_int="3" />
   <AdaptiveDirectionalStratification-DefaultNumberOfSteps                  value_int="2" />
-
+
```
Comment (josephmure) [inline lib/etc/openturns.conf.in:752]: This whitespace is probably unintentional.

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx [hunk 27]
```diff
@@ -79,7 +79,7 @@ public:
 
   /** OptimizationAlgorithm accessor */
   void setNearestPointAlgorithm(const OptimizationAlgorithm & solver);
-
+
```
Comment (josephmure) [inline lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx:82]: Same

File: lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx [hunk 28]
```diff
@@ -103,12 +105,11 @@ protected:
   friend class Factory<Analytical>;
 
 private:
-
   OptimizationAlgorithm nearestPointAlgorithm_;
   RandomVector event_;
   Point physicalStartingPoint_;
   AnalyticalResult result_;
-
+
```
Comment (josephmure) [inline lib/src/Uncertainty/Algorithm/Analytical/openturns/Analytical.hxx:112]: Same (use black on your Python files to automatically remove this kind of things)

File: python/doc/examples/reliability_sensitivity/reliability/plot_axial_stressed_beam.py [hunk 29]
```diff
@@ -142,14 +142,14 @@
 
 # %%
 # We create a NearestPoint algorithm
-myCobyla = ot.Cobyla()
+myalgoOptim = ot.AbdoRackwitz()
```
Comment (josephmure) [inline python/doc/examples/reliability_sensitivity/reliability/plot_axial_stressed_beam.py:145]: The "my" prefix is a bad habit we once had? Since you are renaming the variable, I would suggest "algoOptim" instead of "myalgoOptim"

**Review by josephmure — APPROVED** [review]
Good to go after these small fixes.

File: lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx [hunk 30]
```diff
@@ -131,6 +130,19 @@ void Analytical::run()
   Point value(standardEvent.getImplementation()->getFunction().operator()(origin));
 
   result_.setIsStandardPointOriginInFailureSpace(event_.getOperator().compare(value[0], event_.getThreshold()));
+  
+  /* check if result is valid */
+ const Point physicalSpaceDesignPoint(event_.getImplementation()->getAntecedent().getDistribution().getInverseIsoProbabilisticTransformation().operator()(standardSpaceDesignPoint));
+  
+  const Point valuePhysicalSpaceDesignPoint(event_.getImplementation()->getFunction().operator()(physicalSpaceDesignPoint));
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/Analytical/Analytical.cxx:137]: this should be equal to nearestPointAlgorithm_.getResult().getOptimalValue(), thus avoiding an extra evaluation
